lib/Dialect/AIE/Transforms/AIEAssignBufferDescriptorIDs.cpp

```diff
@@ -0,0 +1,174 @@
+//===- AIEAssignBufferDescriptorIDs.cpp
+//---------------------------------------*- C++
+//-*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 AMD Inc.
+//
+//===----------------------------------------------------------------------===//
+
+#include "aie/Dialect/AIE/IR/AIEDialect.h"
+#include "aie/Dialect/AIE/Transforms/AIEPasses.h"
+
+#include "mlir/Pass/Pass.h"
+
+#define DEBUG_TYPE "aie-assign-bd-ids"
+#define EVEN_BD_ID_START 0
+#define ODD_BD_ID_START 24
+
+using namespace mlir;
+using namespace xilinx;
+using namespace xilinx::AIE;
+
+struct BdIdGenerator {
+  BdIdGenerator(int col, int row, const AIETargetModel &targetModel)
+      : col(col), row(row), isMemTile(targetModel.isMemTile(col, row)) {}
+
+  int32_t nextBdId(int channelIndex) {
+    int32_t bdId = isMemTile && channelIndex & 1 ? oddBdId++ : evenBdId++;
+    while (bdIdAlreadyAssigned(bdId))
+      bdId = isMemTile && channelIndex & 1 ? oddBdId++ : evenBdId++;
+    assignBdId(bdId);
+    return bdId;
+  }
+
+  void assignBdId(int32_t bdId) {
+    assert(!alreadyAssigned.count(bdId) && "bdId has already been assigned");
+    alreadyAssigned.insert(bdId);
+  }
+
+  bool bdIdAlreadyAssigned(int32_t bdId) { return alreadyAssigned.count(bdId); }
+
+  int col;
+  int row;
+  int oddBdId = ODD_BD_ID_START;
+  int evenBdId = EVEN_BD_ID_START;
+  bool isMemTile;
+  std::set<int32_t> alreadyAssigned;
+};
+
+struct AIEAssignBufferDescriptorIDsPass
+    : AIEAssignBufferDescriptorIDsBase<AIEAssignBufferDescriptorIDsPass> {
+  void runOnOperation() override {
+    DeviceOp targetOp = getOperation();
+    const AIETargetModel &targetModel = targetOp.getTargetModel();
+
+    auto memOps = llvm::to_vector_of<TileElement>(targetOp.getOps<MemOp>());
+    llvm::append_range(memOps, targetOp.getOps<MemTileDMAOp>());
+    llvm::append_range(memOps, targetOp.getOps<ShimDMAOp>());
+    for (TileElement memOp : memOps) {
+      int col = memOp.getTileID().col;
+      int row = memOp.getTileID().row;
+      BdIdGenerator gen(col, row, targetModel);
+
+      auto dmaOps = llvm::to_vector_of<DMAOp>(
```

Comment:
The to_vector_of seems unnecessary here?  (yeah, it appears it was in the original code...)

Reply:
Well I put it there... I think I was grabbing `.size()` before but then I moved that check to the verifier.